### ISSUE TITLE
torch.compile produces different fp16 results for where + full_like + cast + add pattern

### ISSUE THREAD

**Issue by ALinrunrun** [pytorch/pytorch#183761]
### 🐛 Describe the bug

`torch.compile` with the Inductor backend produces different results from eager execution for a pure PyTorch function using `torch.where`, `torch.full_like`, fp16 casts, addition, and reduction.

The elementwise fp16 output differs at several positions, and the final float sum is also different.

```
import sys
import torch

x = torch.arange(30, dtype=torch.float32).reshape(2, 3, 5) / 7
m = (torch.arange(30) % 4 == 0).reshape(2, 3, 5)

def f(x, m):
    z = torch.where(m, torch.full_like(x, 0.75), torch.full_like(x, -0.25))
    y = z.to(torch.float16) + x.to(torch.float16)
    return y, y.float().sum()

eager = f(x, m)
compiled = torch.compile(f, backend="inductor", fullgraph=True)(x, m)

diff = (eager[0].float() - compiled[0].float()).abs()

print("input_head:", x.reshape(-1)[:10])
print("mask_head:", m.reshape(-1)[:10])
print("eager_sum:", eager[1])
print("compiled_sum:", compiled[1])
print("max_abs_diff:", diff.max())
print("diff_count:", int((diff != 0).sum()))

sys.exit(0 if (not torch.equal(eager[0], compiled[0]) or not torch.equal(eager[1], compiled[1])) else 1)
```

### Error logs

```
input_head: tensor([0.0000, 0.1429, 0.2857, 0.4286, 0.5714, 0.7143, 0.8571, 1.0000, 1.1429,
        1.2857])
mask_head: tensor([ True, False, False, False,  True, False, False, False,  True, False])
eager_sum: tensor(62.6470)
compiled_sum: tensor(62.6429)
max_abs_diff: tensor(0.0020)
diff_count: 4
```

### Versions

PyTorch version: 2.13.0.dev20260513+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.11.15 (main, Mar 11 2026, 17:20:07) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-20-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration:
GPU 0: NVIDIA RTX 6000 Ada Generation
GPU 1: NVIDIA RTX 6000 Ada Generation

Nvidia driver version: 570.211.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.13.0.dev20260513+cpu
[conda] numpy                      2.4.4                   pypi_0           pypi
[conda] torch                      2.13.0.dev20260513+cpu  pypi_0           pypi

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I attempted to reproduce this issue on PyTorch nightly (`2.13.0.dev20260514+cu130`) on NVIDIA H100. Result: PREDEFINED_MITIGATION.

All 3 runs reproduced the reported numerical mismatch on CPU: eager_sum=62.6470 vs compiled_sum=62.6429, max_abs_diff=0.0020, diff_count=4, elementwise_equal=False, sum_equal=False. Re-running with TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1 eliminated the divergence (max_abs_diff=0.0, both elementwise tensors and the float sum match exactly).

<details>
<summary>Error output</summary>

```
input_head: tensor([0.0000, 0.1429, 0.2857, 0.4286, 0.5714, 0.7143, 0.8571, 1.0000, 1.1429,
        1.2857])
mask_head: tensor([ True, False, False, False,  True, False, False, False,  True, False])
eager_sum: tensor(62.6470)
compiled_sum: tensor(62.6429)
max_abs_diff: tensor(0.0020)
diff_count: 4
elementwise_equal: False
sum_equal: False

ERROR conda.cli.main_run:execute(49): `conda run python /home/karthickps/.repro_agent/workdir/pytorch_pytorch/183761/repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import sys
import torch


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    # The original repro runs on CPU; honor that.
    x = torch.arange(30, dtype=torch.float32).reshape(2, 3, 5) / 7
    m = (torch.arange(30) % 4 == 0).reshape(2, 3, 5)

    def f(x, m):
        z = torch.where(m, torch.full_like(x, 0.75), torch.full_like(x, -0.25))
        y = z.to(torch.float16) + x.to(torch.float16)
        return y, y.float().sum()

    eager = f(x, m)
    compiled = torch.compile(f, backend=backend, fullgraph=True)(x, m)

    elementwise_equal = torch.equal(eager[0], compiled[0])
    sum_equal = torch.equal(eager[1], compiled[1])
    return elementwise_equal and sum_equal


if __name__ == "__main__":
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    x = torch.arange(30, dtype=torch.float32).reshape(2, 3, 5) / 7
    m = (torch.arange(30) % 4 == 0).reshape(2, 3, 5)

    def f(x, m):
        z = torch.where(m, torch.full_like(x, 0.75), torch.full_like(x, -0.25))
        y = z.to(torch.float16) + x.to(torch.float16)
        return y, y.float().sum()

    eager = f(x, m)
    compiled = torch.compile(f, backend=backend, fullgraph=True)(x, m)

    diff = (eager[0].float() - compiled[0].float()).abs()
    print("input_head:", x.reshape(-1)[:10])
    print("mask_head:", m.reshape(-1)[:10])
    print("eager_sum:", eager[1])
    print("compiled_sum:", compiled[1])
    print("max_abs_diff:", diff.max())
    print("diff_count:", int((diff != 0).sum()))
    print("elementwise_equal:", torch.equal(eager[0], compiled[0]))
    print("sum_equal:", torch.equal(eager[1], compiled[1]))

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260514+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260514+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260515+cu130
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260515+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.

**Comment by jansel** [comment]
My agent says this looks like intended Inductor behavior rather than a compiler bug. By default Inductor may elide fp16/bf16 downcast-upcast pairs across fused pointwise ops, so bitwise results can differ from eager for low-precision casts. The exact-eager mode is `TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1`; I confirmed on commit 8dd44388526b4c638900a96f16ecfd1c0b2b0f8d that default Inductor reproduces the reported `62.6470` vs `62.6429` divergence, while the emulate flag makes the elementwise output and sum exactly match eager.
